spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+
+  <div algorithm>
+    To <dfn>obtain a shared storage bottle map</dfn>, given an [=environment settings object=] |environment|, run these steps:
+
+    1. Let |shed| be the [=user agent=]'s [=shared storage shed=].
+    1. Let |shelf| be the result of running [=obtain a shared storage shelf=] with |shed| and |environment|.
+    1. If |shelf| is failure, then return failure.
+    1. Let |bucket| be |shelf|'s [=bucket map=]["`default`"].
+    1. Let |bottle| be |bucket|'s [=bottle map=]["`sharedStorage`"].
+    1. Let |proxyMap| be a new [=storage proxy map=] whose [=backing map=] is |bottle|'s [=map=].
+    1. [=set/Append=] |proxyMap| to |bottle|'s [=proxy map reference set=].
+    1. Return |proxyMap|.
+  </div>
+
+  ## The [=Shared Storage Database=] ## {#database}
+
+  A [=/browsing context=] has an associated <dfn>shared storage database</dfn>, which provides methods to [=shared storage database/store an entry in the database|store=], [=shared storage database/retrieve an entry from the database|retrieve=], [=shared storage database/delete an entry from the database|delete=], [=shared storage database/clear all entries in the database|clear=], and [=shared storage database/purge expired entries from the database|purge expired=] data, and additional methods as below. The data in the [=shared storage database|database=] take the form of [=shared storage database/entry|entries=].
+
+  Each [=shared storage database=] has a <dfn for="shared storage database">shared storage database queue</dfn>, which is a [=parallel queue=] on which the each of the [=shared storage database=]'s methods run when calls to them are initiated from that [=/browsing context=].
+
+  Each <dfn for="shared storage database">entry</dfn> consists of a [=entry/key=] and a [=entry/value struct=].
+
+  An [=shared storage database/entry=]'s <dfn for=entry>key</dfn> is a [=string=].
+
+  [=User agents=] may specify the <dfn for=key>maximum length</dfn> of a [=entry/key=].
+
+  Since [=entry/keys=] are used to organize and efficiently retrieve [=shared storage database/entry|entries=], [=entry/keys=] must appear at most once in any given [=shared storage database=].
+
+  An [=shared storage database/entry=]'s <dfn for=entry>value struct</dfn> is a [=struct=] composed of [=string=] <dfn for="value struct">value</dfn> and {{DOMHighResTimeStamp}} <dfn for="value struct">last updated</dfn> (from the [=Unix Epoch=]).
+
+  Note: Some Shared Storage implementations may use SQLite implementations that store timestamps as 64-bit unsigned integers rather than doubles, thereby making it necessary to use legacy {{EpochTimeStamp}}s instead of {{DOMHighResTimeStamp}}s.
+
+  <span class=todo>In Chrome, for the reason given in the note above, we use {{EpochTimeStamp}}s, which are discouraged. How can we reconcile this?</span>
```

Comment:
Can you explain this a bit more?  Surely sqlite implementations can store double values.  What prevents us from accurately representing a DOMHighResTimeStamp here?  Can we not convert whatever our internal representation is into a DOMHighResTimeStamp when its exposed to script?

Reply:
Here is the code reference:
https://source.chromium.org/chromium/chromium/src/+/main:sql/statement.cc;drc=c85c710dec58dfdcf2f35af647a0a5155639edd5;l=469

It would probably not be impossible to roll our own version using a double, but I don't think that's best practice.

But sure, we could convert our int to a double when it's exposed to script.

Reply:
Yea, I think we should write the spec without these kinds of implementation conversion details.  It shouldn't really mention sqlite as a possible implementation database under the hood.  If there is any conversion needed we should just do it in our code and not bring it up in the spec.